From 17489962a4bd5a845ad5029a7aaab425fed1e323 Mon Sep 17 00:00:00 2001
From: Harish Rachakonda <quic_rachakon@quicinc.com>
Date: Mon, 22 Jul 2024 15:26:10 +0530
Subject: [PATCH] iw: configure the antenna gain

Configured antenna gain. Used to reduce
transmit power to stay within regulatory limits

Signed-off-by: Harish Rachakonda <quic_rachakon@quicinc.com>
Signed-off-by: Rajat Soni <quic_rajson@quicinc.com>
---
 phy.c | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/phy.c b/phy.c
index 40d4d8e..59ad0d5 100644
--- a/phy.c
+++ b/phy.c
@@ -899,3 +899,30 @@ static int handle_get_txq(struct nl80211_state *state,
 COMMAND(get, txq, "",
 	NL80211_CMD_GET_WIPHY, 0, CIB_PHY, handle_get_txq,
 	"Get TXQ parameters.");
+
+static int handle_antenna_gain(struct nl80211_state *state,
+			       struct nl_msg *msg,
+			       int argc, char **argv,
+			       enum id_input id)
+{
+	char *endptr;
+	int dbm;
+
+	/* get the required args */
+	if (argc != 1)
+		return 1;
+
+	dbm = strtol(argv[0], &endptr, 10);
+	if (*endptr)
+		return 2;
+
+	NLA_PUT_U32(msg, NL80211_ATTR_WIPHY_ANTENNA_GAIN, dbm);
+
+	return 0;
+
+ nla_put_failure:
+	return -ENOBUFS;
+}
+COMMAND(set, antenna_gain, "<antenna gain in dBm>",
+	NL80211_CMD_SET_WIPHY, 0, CIB_PHY, handle_antenna_gain,
+	"Specify antenna gain.");
-- 
2.34.1

